src/observer.js

```diff
@@ -138,26 +138,104 @@ const reactiveMixin = {
         let isForcingUpdate = false
 
         function makePropertyObservableReference(propName) {
-            let valueHolder = this[propName]
-            const atom = new Atom("reactive " + propName)
+            let currentValue = this[propName]
+            let currentValueKeys = null == currentValue ? [] : Object.keys(currentValue)
+            const changeKeysAtom = new Atom("this." + propName + " [change keys]")
+            const individualAtoms = {}
+            const individualValues = {}
+
+            /**
+             * In-place convert properties to getter+setter
+             */
+            function convertStorageToReactive(storage) {
+                if (null == storage) {
+                    return storage
+                }
+                // props object is frozen, so we need to clone it
+                // we can do it safely inside componentWillMount (react won't warn us about it)
+                // btw, state object can not be cloned here, (react warns about direct state modification), however
+                // state object is not frozen, so we do not need to clone it to instrument
+                if (Object.isFrozen && Object.isFrozen(storage)) {
+                    storage = Object.assign({}, storage)
+                }
+                Object.keys(storage).forEach(function(key) {
+                    if (!(key in individualAtoms)) {
+                        individualAtoms[key] = new Atom("this." + propName + "." + key)
+                    }
+                    const currentKeyValue = storage[key]
+                    delete storage[key]
+                    Object.defineProperty(storage, key, {
+                        configurable: true,
+                        enumerable: true,
+                        get: function get() {
+                            individualAtoms[key].reportObserved()
+                            return individualValues[key]
+                        },
+                        set: function set(v) {
```

Comment:
I think we can make property read-only (by defining only `get`), this can help to make props read-only too (BC with current react behaviour)